### ISSUE TITLE
Script failed with "Chrome For Testing" tied to ChromeDriver

### ISSUE THREAD

**Issue by profucius** [RileyXX/IMDB-Trakt-Syncer#116]
### Is there already an issue for your problem?

- [X] I have checked older issues, open and closed

### Bug Description

This is an extension of [my comment over here](https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/16#issuecomment-2545880832), the following issue arose after doing what I wrote there, the script then said IMDB was  "not logged in", so I followed the instructions on issue 2 to manually login. After doing that I ran the script again, and the below happened.

---

```
C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer>IMDBTraktSyncer
Starting IMDBTraktSyncer....
Your settings are saved at:
C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer
Starting WebDriver...

--- ERROR ---
An error occurred while running the script.
Please submit the error to GitHub with the following information:
--------------------------------------------------
Submit the error here: https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/new?template=bug_report.yml
--------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 104, in main
    sign_in_link.click()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webelement.py", line 94, in click
    self._execute(Command.CLICK_ELEMENT)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webelement.py", line 395, in _execute
    return self._parent.execute(command, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: disconnected: Unable to receive message from renderer
  (failed to check if window was closed: disconnected: not connected to DevTools)
  (Session info: chrome=131.0.6778.108)
Stacktrace:
        GetHandleVerifier [0x003E34A3+25059]
        (No symbol) [0x0036CEA4]
        (No symbol) [0x0024BEC3]
        (No symbol) [0x0023D773]
        (No symbol) [0x0023D481]
        (No symbol) [0x0023B91B]
        (No symbol) [0x0023BF8D]
        (No symbol) [0x002512A2]
        (No symbol) [0x002538EA]
        (No symbol) [0x00253967]
        (No symbol) [0x0029396E]
        (No symbol) [0x00291DAD]
        (No symbol) [0x002910AF]
        (No symbol) [0x00286169]
        (No symbol) [0x002B1EFC]
        (No symbol) [0x00285A24]
        (No symbol) [0x002B2194]
        (No symbol) [0x002CB51E]
        (No symbol) [0x002B1C96]
        (No symbol) [0x00283FAC]
        (No symbol) [0x00284F3D]
        GetHandleVerifier [0x006D5613+3113811]
        GetHandleVerifier [0x006EA2DA+3199002]
        GetHandleVerifier [0x006E2AB2+3168242]
        GetHandleVerifier [0x00483310+680016]
        (No symbol) [0x003757ED]
        (No symbol) [0x00372A98]
        (No symbol) [0x00372C35]
        (No symbol) [0x00365890]
        BaseThreadInitThunk [0x75FC5D49+25]
        RtlInitializeExceptionChain [0x7787CEBB+107]
        RtlGetAppContainerNamedObjectPath [0x7787CE41+561]


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Scripts\IMDBTraktSyncer.exe\__main__.py", line 7, in <module>
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 780, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: disconnected: not connected to DevTools
  (failed to check if window was closed: disconnected: not connected to DevTools)
  (Session info: chrome=131.0.6778.108)
Stacktrace:
        GetHandleVerifier [0x003E34A3+25059]
        (No symbol) [0x0036CEA4]
        (No symbol) [0x0024BEC3]
        (No symbol) [0x0023B21E]
        (No symbol) [0x0023BF8D]
        (No symbol) [0x0024C3F2]
        (No symbol) [0x0022D555]
        (No symbol) [0x002BDAEE]
        (No symbol) [0x002B1EC3]
        (No symbol) [0x00283FAC]
        (No symbol) [0x00284F3D]
        GetHandleVerifier [0x006D5613+3113811]
        GetHandleVerifier [0x006EA2DA+3199002]
        GetHandleVerifier [0x006E2AB2+3168242]
        GetHandleVerifier [0x00483310+680016]
        (No symbol) [0x003757ED]
        (No symbol) [0x00372A98]
        (No symbol) [0x00372C35]
        (No symbol) [0x00365890]
        BaseThreadInitThunk [0x75FC5D49+25]
        RtlInitializeExceptionChain [0x7787CEBB+107]
        RtlGetAppContainerNamedObjectPath [0x7787CE41+561]
```

### Environment

OS: Windows 11
Python: 3.12
Chrome For Testing & ChromeDriver (both): 131.0.6778.108

**Comment by profucius** [comment]
Update: I just realized that you recently updated the script to no longer need ChromeDriver, in lieu of WebDriver. After uninstalling ChromeDriver, the script now says "Successfully signed in to IMDB". However it seems to be hung there, is there any way I can see if it is doing anything?

Suggestion: Update the script to block accessing the old ChromeDriver if it is still installed.

**Comment by RileyXX** [comment]
Hmm from the error message it looks like chrome seems to either be crashing or not launching it all. It's odd there isn't more error messages though. The error provided seems to be triggered around the same time in the script where the first page is loaded (sign in). It's basically is saying the sign in element wasn't found, due to chromedriver being unable to connect with chrome webtools (chrome not running/crashed).

Whether you still have chromedriver-py installed or not, it shouldn't effect the script in anyway because it's not being called in the latest version.

The latest version using webdriver-manager is basically a download manager for chromedriver. I think the problem might be with webdriver-manager not being able to determine which chrome version you are using since you are using portable chrome, but it's odd that it isn't giving any error messages specific to that.

You might need to try my suggestion from the other thread, modify `IMDBTraktSyncer.py` in a text editor and replace the line `service = webdriver.chrome.service.Service(ChromeDriverManager().install(), popen_kw=popen_kw)` with service = `webdriver.chrome.service.Service("path/to/chrome.exe")` and see if that changes anything. 

Other thing you could try, based on documentation here https://pypi.org/project/webdriver-manager/ try forcing the download of the chromedriver version you need. You'll need to modify `IMDBTraktSyncer.py` in a text editor and replace the line `service = webdriver.chrome.service.Service(ChromeDriverManager().install(), popen_kw=popen_kw)` with `service = webdriver.chrome.service.Service(ChromeDriverManager(driver_version="131.0.6778.108").install(), popen_kw=popen_kw)` in your case. 

Give those solutions a try and let me know what happens.

**Comment by rufftruffles** [comment]
@RileyXX almost same issue as the poster above but on linux:

`$ pyimdb/bin/IMDBTraktSyncer                     Starting IMDBTraktSyncer....
Your settings are saved at:
/home/tor/pyimdb/lib/python3.11/site-packages/IMDBTraktSyncer
Starting WebDriver...
Error initializing WebDriver: Message: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 126.0.6478.62 with binary path /home/tor/.chrome/chrome
Stacktrace:
#0 0x5640fc9294e3 <unknown>
#1 0x5640fc658c76 <unknown>
#2 0x5640fc68604a <unknown>
#3 0x5640fc6814a1 <unknown>
#4 0x5640fc67e029 <unknown>
#5 0x5640fc6bcccc <unknown>
#6 0x5640fc6bc47f <unknown>
#7 0x5640fc6b3de3 <unknown>
#8 0x5640fc6892dd <unknown>
#9 0x5640fc68a34e <unknown>
#10 0x5640fc8e93e4 <unknown>
#11 0x5640fc8ed3d7 <unknown>
#12 0x5640fc8f7b20 <unknown>
#13 0x5640fc8ee023 <unknown>
#14 0x5640fc8bc1aa <unknown>
#15 0x5640fc9126b8 <unknown>
#16 0x5640fc912847 <unknown>
#17 0x5640fc922243 <unknown>
#18 0x7f77c6e10144 <unknown> `

**Comment by RileyXX** [comment]
@rufftruffles It looks like the wrong chromedriver version is being used. Can you confirm how you have Chrome installed? Are you using "[Chrome for Testing](https://googlechromelabs.github.io/chrome-for-testing/)" portable version? If so then this is not officially supported yet with this script. You would need to try some of the solutions in my [above](https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/116#issuecomment-2547003105) comment.

**Comment by RileyXX** [comment]
I'm working on a version now that downloads and manages [Chrome for Testing](https://googlechromelabs.github.io/chrome-for-testing/) (chrome portable) automatically, which would remove the Chrome install requirement. 

However, I ran into a bug which I believe to be specific to Chromium's "Chrome for Testing" branch. Essentially the browser is not finishing loading pages after downloading a file (the csv exports from IMDB). The script reads this as a page timeout error and retries until eventually failing with the message `Unable to receive message from renderer`. 

In regards to OP @profucius, the solutions in my [above comment](https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/116#issuecomment-2547003105) might get you through that initial error, but would probably hang up at the same spot as me during my testing. That being said, the bug might be specific to the Windows version of "Chrome for Testing", so it's still worth trying the solutions in my [above comment](https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/116#issuecomment-2547003105) anyway to see what happens.

So atleast for now, I don't think it will be possible to run this script using Chrome for Testing. There may be some possible workarounds for it. I'd rather wait for a fix and do it correctly. I'm still looking into this, and I submitted a bug report on the Chromium dev forum so I'm waiting to hear back on that. 

GIF capture of Chrome for Testing bug I ran into (page loading endlessly after download):
![12-31-2024_683](https://github.com/user-attachments/assets/c0e6e8f6-16d1-4ed4-9914-00352dcc9686)

**Comment by RileyXX** [comment]
So I've been doing some more testing, it looks like there is a [bug](https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/116#issuecomment-2566890102) specific to the "Chrome for Testing" browser. I reported it, and the Chrome dev team is investigating. However, I tested the script using the "chrome-headless-shell" browser and it worked without issues. So I should be able to proceed developing this new version which would eliminate the Chrome install requirement. It should solve both of your issues and also make the script more portable for future docker implementation support.

**Comment by RileyXX** [comment]
I just pushed a new update [v3.1.0](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/v3.1.0) which automatically downloads and manages the "Chrome for Testing" headless shell browser. I think this might solve both issues you guys were having. 

Give it a try and let me know how it goes.
````
python -m pip install IMDBTraktSyncer --upgrade
````

**Comment by profucius** [comment]
Hey, so I gave it a try this morning. I seem to be having an issue with it, there is an error it gives me. Pasting the cmd log output below:

```
D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer
Do you want to sync watchlists? (y/n)
Enter your choice: y
Do you want to sync ratings? (y/n)
Enter your choice: n
Movies and Episodes are removed from watchlists after 1 play.
Shows are removed when atleast 80% of the episodes are watched AND the series is marked as ended or cancelled.
Do you want to remove watched items watchlists? (y/n)
Enter your choice: y
Please note: reviews synced to IMDB will use "My Review" as the title field.
Do you want to sync reviews? (y/n)
Enter your choice: n
Starting WebDriver...
Error initializing WebDriver: Message: session not created: This version of ChromeDriver only supports Chrome version 128
Current browser version is 131.0.6778.264 with binary path D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\131.0.6778.264\chrome-headless-shell-win64\chrome-headless-shell.exe
Stacktrace:
        GetHandleVerifier [0x00007FF63CBD9412+29090]
        (No symbol) [0x00007FF63CB4E239]
        (No symbol) [0x00007FF63CA0B1DA]
        (No symbol) [0x00007FF63CA4D6D2]
        (No symbol) [0x00007FF63CA4C771]
        (No symbol) [0x00007FF63CA472AC]
        (No symbol) [0x00007FF63CA41E1A]
        (No symbol) [0x00007FF63CA9233D]
        (No symbol) [0x00007FF63CA919F0]
        (No symbol) [0x00007FF63CA86493]
        (No symbol) [0x00007FF63CA509D1]
        (No symbol) [0x00007FF63CA51B31]
        GetHandleVerifier [0x00007FF63CEF871D+3302573]
        GetHandleVerifier [0x00007FF63CF44243+3612627]
        GetHandleVerifier [0x00007FF63CF3A417+3572135]
        GetHandleVerifier [0x00007FF63CC95EB6+801862]
        (No symbol) [0x00007FF63CB5945F]
        (No symbol) [0x00007FF63CB54FB4]
        (No symbol) [0x00007FF63CB55140]
        (No symbol) [0x00007FF63CB4461F]
        BaseThreadInitThunk [0x00007FFACCE7DBE7+23]
        RtlUserThreadStart [0x00007FFACD805A4C+44]
```

**Comment by profucius** [comment]
Just tried your advise in this order:
- I deleted the symlinks I made in `C:\Users\Admin\AppData\Local\Google` (actually just renamed the whole folder).
- Checked the `site-packages\IMDBTraktSyncer` dir and I see a `Chrome` folder within that has a `131.0.6778.264` subfolder which has a `chrome-headless-shell-win64` subfolder. This seems like what the script wants to use.
- Checked the `C:\Users\Admin\.cache\selenium\chromedriver\win64` folder, deleted the old `128.0.6613.137` subfolder.
- Ran IMDBTraktSyncer.py
- Checked the `C:\Users\Admin\.cache\selenium\chromedriver\win64` folder, and the `128.0.6613.137` subfolder has reappeared.

I tried all the above again, but this time I deleted the`site-packages\IMDBTraktSyncer\Chrome` folder. Ran the script again, and it redownloaded the chromedriver again. But the `.cache\selenium` folder version was `128.0.6613.137` again.

I then tried manually replacing the `chromedriver.exe` with v131, and finally the script made some progress. Now it says I am not signed in to IMDb.

Two questions: It seems odd that this would be getting v128, any idea why? And, how do I sign in to IMDb if it's using this new headless version?

**Comment by RileyXX** [comment]
> And, how do I sign in to IMDb if it's using this new headless version?

That's good you got to the not signed in, sounds like it's working with your manual fix. Captcha trigger is common issue. 

Just follow the instructions in the command prompt. Basically you need to open your browser (preferably chrome and from same computer), login/out of imdb until you trigger captcha and then do it until the captcha is no longer triggered. I found using incognito tab works the best for this. If you run the script periodically it usually doesn't trigger the captcha check. And when it does, it resolves on its own eventually (if you're planning on running it in the background).

**Comment by RileyXX** [comment]
Some more things you can try to fix the wrong driver download:

Another project might have set the selenium browser/chromedriver version in your systems environment labels. Try running these 2 lines in command prompt to reset the environment labels for selenium:

```
set SE_BROWSER_VERSION=
set SE_CHROME_VERSION=
```

Next, in command prompt cd to the selenium-manager.exe directory (replace username and python version fields):

```
cd C:\Users\YOUR_USERNAME\AppData\Local\Programs\Python\YOUR_PYTHON_VERSION\Lib\site-packages\selenium\webdriver\common\windows
```
Once you're there, in command prompt run this line:

```
selenium-manager.exe --clear-cache --driver chromedriver --browser chrome
```
Then run the script again and see if that helps.

**Comment by RileyXX** [comment]
This is the documentation for selenium manager, and relevant commands: https://www.selenium.dev/documentation/selenium_manager/

I'm still learning more about it myself as it's still relatively new feature. I'm thinking about making some changes and using selenium manager entirely for this project to manager chrome and chromedriver downloads. That might help with inconsistencies like in your case. Either that, or I can handle chrome + chromedriver downloads automatically in the script and apply the path directly which would avoid any selenium manager issues. I would prefer to use selenium manager though since it's a cleaner approach.

**Comment by profucius** [comment]
Sounds good, thank you for all the helpful advise and documentation. I'll certainly dig deeper into it soon, this weekend should have some time! I'll report back here if there are any further issues, but if I at least got the IMDb sign-in message then yeah I think it should work.

**Comment by RileyXX** [comment]
So I decided to handle chromedriver + chrome browser downloads internally, instead of relying on selenium's built in tools or other third party tools. This should give more control over driver mis matches. 

I think [v3.2.2](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/v3.2.2) or later should fix the issue you were having. @profucius 

````bash
python -m pip install IMDBTraktSyncer --upgrade
````
Give it a try and let me know how it goes. I'm going to close this thread as completed, if you still have issues let me know and I can reopen it.

**Comment by profucius** [comment]
Hey I just tried your newest 3.2.7 version, and I'm having a new issue. I think it's a great idea to bundle it all together as much as possible, I agree it should reduce issues going forward.

```
C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer>IMDBTraktSyncer.py
Starting IMDBTraktSyncer....
Your settings are saved at:
D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer
Starting WebDriver...
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\AutofillStates: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\CertificateRevocation: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\Crashpad: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\Crowd Deny: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\Default: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\FileTypePolicies: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\Fingerprinting Protection Filter: name 'stat' is not defined
...
many more lines just like this
...
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\ShaderCache\data_3: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\ShaderCache\index: name 'stat' is not defined
Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\Subresource Filter\Unindexed Rules: name 'stat' is not defined
   - Selenium WebDriver Error: Message: unknown error: net::ERR_CONNECTION_CLOSED URL: https://www.imdb.com/ (File: errorHandling.py, Line: 229)
   - Retryable network error detected: Message: unknown error: net::ERR_CONNECTION_CLOSED Retrying... Time Remaining: 180s

--- ERROR ---
An error occurred while running the script.
Please submit the error to GitHub with the following information:
--------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 146, in main
    element = wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, ".nav__userMenu.navbar__user")))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\support\wait.py", line 96, in until
    value = method(self._driver)
            ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\support\expected_conditions.py", line 84, in _predicate
    return driver.find_element(*locator)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 770, in find_element
    return self.execute(Command.FIND_ELEMENT, {"using": by, "value": value})["value"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id: session deleted as the browser has closed the connection
from disconnected: not connected to DevTools
  (Session info: chrome=132.0.6834.83)
Stacktrace:
        GetHandleVerifier [0x00007FF7C200FBE5+28741]
        (No symbol) [0x00007FF7C1F77890]
        (No symbol) [0x00007FF7C1E14FDA]
        (No symbol) [0x00007FF7C1E01B0C]
        (No symbol) [0x00007FF7C1E019D0]
        (No symbol) [0x00007FF7C1E179F1]
        (No symbol) [0x00007FF7C1EAF9AA]
        (No symbol) [0x00007FF7C1E8FCB3]
        (No symbol) [0x00007FF7C1E59FB3]
        (No symbol) [0x00007FF7C1E5B331]
        GetHandleVerifier [0x00007FF7C234A73D+3414941]
        GetHandleVerifier [0x00007FF7C235E64A+3496618]
        GetHandleVerifier [0x00007FF7C235413D+3454365]
        GetHandleVerifier [0x00007FF7C20D848B+850155]
        (No symbol) [0x00007FF7C1F837FF]
        (No symbol) [0x00007FF7C1F7F0C4]
        (No symbol) [0x00007FF7C1F7F25D]
        (No symbol) [0x00007FF7C1F6E079]
        BaseThreadInitThunk [0x00007FFEC048DBE7+23]
        RtlUserThreadStart [0x00007FFEC1B85A4C+44]


--------------------------------------------------
Submit the error here: https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/new?template=bug_report.yml
--------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 146, in main
    element = wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, ".nav__userMenu.navbar__user")))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\support\wait.py", line 96, in until
    value = method(self._driver)
            ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\support\expected_conditions.py", line 84, in _predicate
    return driver.find_element(*locator)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 770, in find_element
    return self.execute(Command.FIND_ELEMENT, {"using": by, "value": value})["value"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id: session deleted as the browser has closed the connection
from disconnected: not connected to DevTools
  (Session info: chrome=132.0.6834.83)
Stacktrace:
        GetHandleVerifier [0x00007FF7C200FBE5+28741]
        (No symbol) [0x00007FF7C1F77890]
        (No symbol) [0x00007FF7C1E14FDA]
        (No symbol) [0x00007FF7C1E01B0C]
        (No symbol) [0x00007FF7C1E019D0]
        (No symbol) [0x00007FF7C1E179F1]
        (No symbol) [0x00007FF7C1EAF9AA]
        (No symbol) [0x00007FF7C1E8FCB3]
        (No symbol) [0x00007FF7C1E59FB3]
        (No symbol) [0x00007FF7C1E5B331]
        GetHandleVerifier [0x00007FF7C234A73D+3414941]
        GetHandleVerifier [0x00007FF7C235E64A+3496618]
        GetHandleVerifier [0x00007FF7C235413D+3454365]
        GetHandleVerifier [0x00007FF7C20D848B+850155]
        (No symbol) [0x00007FF7C1F837FF]
        (No symbol) [0x00007FF7C1F7F0C4]
        (No symbol) [0x00007FF7C1F7F25D]
        (No symbol) [0x00007FF7C1F6E079]
        BaseThreadInitThunk [0x00007FFEC048DBE7+23]
        RtlUserThreadStart [0x00007FFEC1B85A4C+44]


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 845, in <module>
    main()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 839, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id
Stacktrace:
        GetHandleVerifier [0x00007FF7C200FBE5+28741]
        (No symbol) [0x00007FF7C1F77890]
        (No symbol) [0x00007FF7C1E14E09]
        (No symbol) [0x00007FF7C1E590E4]
        (No symbol) [0x00007FF7C1E8FDB9]
        (No symbol) [0x00007FF7C1E8A76B]
        (No symbol) [0x00007FF7C1E89836]
        (No symbol) [0x00007FF7C1DE2EC5]
        GetHandleVerifier [0x00007FF7C234A73D+3414941]
        GetHandleVerifier [0x00007FF7C235E64A+3496618]
        GetHandleVerifier [0x00007FF7C235413D+3454365]
        GetHandleVerifier [0x00007FF7C20D848B+850155]
        (No symbol) [0x00007FF7C1F837FF]
        (No symbol) [0x00007FF7C1DE1AD4]
        GetHandleVerifier [0x00007FF7C23C5198+3917304]
        BaseThreadInitThunk [0x00007FFEC048DBE7+23]
        RtlUserThreadStart [0x00007FFEC1B85A4C+44]
```

**Comment by RileyXX** [comment]
Thanks for the log. Appreciate your help troubleshooting. The error: `Error modifying permissions for D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\Chrome\132.0.6834.83\chrome-win64\userData\AutofillStates: name 'stat' is not defined` is due to a missing import bug. 

The subsequent errors seems to be due to a connection error when loading the IMDB homepage in chrome `Retryable network error detected: Message: unknown error: net::ERR_CONNECTION_CLOSED Retrying... Time Remaining: 180s`. It looks like, based on your log it did not retry the retryable error, thus failing due to connection issue. It's also possible this error was caused by a permission issue from the chrome browser extracted files, manual exit of the script or some other unknown issue.

In any case, both of these problems should be fixed in [v3.2.8](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/v3.2.8).

```bash
python -m pip install IMDBTraktSyncer --upgrade
```

Give it a test and let me know how it goes.

**Comment by profucius** [comment]
You're so on top of this! Okay the 3.2.8 update fixed the above issue. I got the "IMDB not signed in" message this time. I followed the instructions and logged into IMDB. The script is now processing IMDB and Trakt data. It took a few minutes on Processing IMDB but finally proceeded.

I got a couple different errors. I'll just paste them here since we're talking.

```
Starting WebDriver...
Successfully signed in to IMDB
Processing Trakt Data
Processing Trakt Data Complete
Processing IMDB Data
Exception occurred while getting IMDB reviews: <class 'selenium.common.exceptions.InvalidSessionIdException'>
```
then it proceeded to add items to trakt watchlist.
then it proceeded to add items to imdb watchlist
It seems to have failed on all 150 items it tried to do, all lines similar to this:
```
 - Adding item (1 of 150): Solo Leveling (2024) to IMDB Watchlist (tt21209876)
   - Selenium WebDriver Error: Message: invalid session id URL: https://www.imdb.com/title/tt21209876/ (File: errorHandling.py, Line: 221)
   - Non-retryable WebDriver error encountered. Aborting.
   - Failed to add item (1 of 150): Solo Leveling (2024) to IMDB Watchlist (tt21209876)
```
the same as above when it reached the Removing Watched Items from IMDB Watchlist section, for 4 of 4 items.
then it closed Webdriver. This is the bottom of the log before it closed out.

```
Closing webdriver...

--- ERROR ---
An error occurred while running the script.
Please submit the error to GitHub with the following information:
--------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 827, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id
Stacktrace:
        GetHandleVerifier [0x00007FF7C200FBE5+28741]
        (No symbol) [0x00007FF7C1F77890]
        (No symbol) [0x00007FF7C1E14E09]
        (No symbol) [0x00007FF7C1E590E4]
        (No symbol) [0x00007FF7C1E8FDB9]
        (No symbol) [0x00007FF7C1E8A76B]
        (No symbol) [0x00007FF7C1E89836]
        (No symbol) [0x00007FF7C1DE2EC5]
        GetHandleVerifier [0x00007FF7C234A73D+3414941]
        GetHandleVerifier [0x00007FF7C235E64A+3496618]
        GetHandleVerifier [0x00007FF7C235413D+3454365]
        GetHandleVerifier [0x00007FF7C20D848B+850155]
        (No symbol) [0x00007FF7C1F837FF]
        (No symbol) [0x00007FF7C1DE1AD4]
        GetHandleVerifier [0x00007FF7C23C5198+3917304]
        BaseThreadInitThunk [0x00007FFEC048DBE7+23]
        RtlUserThreadStart [0x00007FFEC1B85A4C+44]


--------------------------------------------------
Submit the error here: https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/new?template=bug_report.yml
--------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 827, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id
Stacktrace:
        GetHandleVerifier [0x00007FF7C200FBE5+28741]
        (No symbol) [0x00007FF7C1F77890]
        (No symbol) [0x00007FF7C1E14E09]
        (No symbol) [0x00007FF7C1E590E4]
        (No symbol) [0x00007FF7C1E8FDB9]
        (No symbol) [0x00007FF7C1E8A76B]
        (No symbol) [0x00007FF7C1E89836]
        (No symbol) [0x00007FF7C1DE2EC5]
        GetHandleVerifier [0x00007FF7C234A73D+3414941]
        GetHandleVerifier [0x00007FF7C235E64A+3496618]
        GetHandleVerifier [0x00007FF7C235413D+3454365]
        GetHandleVerifier [0x00007FF7C20D848B+850155]
        (No symbol) [0x00007FF7C1F837FF]
        (No symbol) [0x00007FF7C1DE1AD4]
        GetHandleVerifier [0x00007FF7C23C5198+3917304]
        BaseThreadInitThunk [0x00007FFEC048DBE7+23]
        RtlUserThreadStart [0x00007FFEC1B85A4C+44]


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 845, in <module>
    main()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 839, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id
Stacktrace:
        GetHandleVerifier [0x00007FF7C200FBE5+28741]
        (No symbol) [0x00007FF7C1F77890]
        (No symbol) [0x00007FF7C1E14E09]
        (No symbol) [0x00007FF7C1E590E4]
        (No symbol) [0x00007FF7C1E8FDB9]
        (No symbol) [0x00007FF7C1E8A76B]
        (No symbol) [0x00007FF7C1E89836]
        (No symbol) [0x00007FF7C1DE2EC5]
        GetHandleVerifier [0x00007FF7C234A73D+3414941]
        GetHandleVerifier [0x00007FF7C235E64A+3496618]
        GetHandleVerifier [0x00007FF7C235413D+3454365]
        GetHandleVerifier [0x00007FF7C20D848B+850155]
        (No symbol) [0x00007FF7C1F837FF]
        (No symbol) [0x00007FF7C1DE1AD4]
        GetHandleVerifier [0x00007FF7C23C5198+3917304]
        BaseThreadInitThunk [0x00007FFEC048DBE7+23]
        RtlUserThreadStart [0x00007FFEC1B85A4C+44]
```

**Comment by profucius** [comment]
I also noticed that the script did not ask the four questions at the beginning that it used to ask. The Y/N if you want to sync one way or the other, etc.

**Comment by RileyXX** [comment]
> I also noticed that the script did not ask the four questions at the beginning that it used to ask. The Y/N if you want to sync one way or the other, etc.

This is expected behavior. Credentials should only need to be entered once on first run. Credentials are carried over when upgrading. You can reset your entered credentials using `IMDBTraktSyncer --clear-user-data` should you ever need to change your settings/credentials.

With this error`'selenium.common.exceptions.InvalidSessionIdException'` the driver is trying to reference an invalid session (browser closed prematurely), wrong chrome tab, etc. I think this might be related to a temporary fix for a [chromium bug](https://github.com/RileyXX/IMDB-Trakt-Syncer/issues/121#issuecomment-2577513043). 

I made a few changes in [v3.2.9](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/v3.2.9) that might fix this problem.

```bash
python -m pip install IMDBTraktSyncer --upgrade
```

If you're still having issues, let me know and I can try a some other possible solutions.

**Comment by profucius** [comment]
Tried 3.3.0, unfortunately the same issue occurred. I'll paste the log output, in case it happens to be different than before:

```
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 396, in main
    driver.execute_script("arguments[0].click();", watchlist_button)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 444, in execute_script
    return self.execute(command, {"script": script, "args": converted_args})["value"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id: session deleted as the browser has closed the connection
from disconnected: not connected to DevTools
  (Session info: chrome=132.0.6834.83)
Stacktrace:
	GetHandleVerifier [0x00007FF7C200FBE5+28741]
	(No symbol) [0x00007FF7C1F77890]
	(No symbol) [0x00007FF7C1E14FDA]
	(No symbol) [0x00007FF7C1E01B0C]
	(No symbol) [0x00007FF7C1E019D0]
	(No symbol) [0x00007FF7C1E179F1]
	(No symbol) [0x00007FF7C1EAF9AA]
	(No symbol) [0x00007FF7C1E8FCB3]
	(No symbol) [0x00007FF7C1E59FB3]
	(No symbol) [0x00007FF7C1E5B331]
	GetHandleVerifier [0x00007FF7C234A73D+3414941]
	GetHandleVerifier [0x00007FF7C235E64A+3496618]
	GetHandleVerifier [0x00007FF7C235413D+3454365]
	GetHandleVerifier [0x00007FF7C20D848B+850155]
	(No symbol) [0x00007FF7C1F837FF]
	(No symbol) [0x00007FF7C1F7F0C4]
	(No symbol) [0x00007FF7C1F7F25D]
	(No symbol) [0x00007FF7C1F6E079]
	BaseThreadInitThunk [0x00007FFEC048DBE7+23]
	RtlUserThreadStart [0x00007FFEC1B85A4C+44]


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Scripts\IMDBTraktSyncer.exe\__main__.py", line 7, in <module>
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 839, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id
Stacktrace:
	GetHandleVerifier [0x00007FF7C200FBE5+28741]
	(No symbol) [0x00007FF7C1F77890]
	(No symbol) [0x00007FF7C1E14E09]
	(No symbol) [0x00007FF7C1E590E4]
	(No symbol) [0x00007FF7C1E8FDB9]
	(No symbol) [0x00007FF7C1E8A76B]
	(No symbol) [0x00007FF7C1E89836]
	(No symbol) [0x00007FF7C1DE2EC5]
	GetHandleVerifier [0x00007FF7C234A73D+3414941]
	GetHandleVerifier [0x00007FF7C235E64A+3496618]
	GetHandleVerifier [0x00007FF7C235413D+3454365]
	GetHandleVerifier [0x00007FF7C20D848B+850155]
	(No symbol) [0x00007FF7C1F837FF]
	(No symbol) [0x00007FF7C1DE1AD4]
	GetHandleVerifier [0x00007FF7C23C5198+3917304]
	BaseThreadInitThunk [0x00007FFEC048DBE7+23]
	RtlUserThreadStart [0x00007FFEC1B85A4C+44]
```

**Comment by RileyXX** [comment]
I made a few more improvements with driver handling in [v3.3.1](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/v3.3.1).

Give it a try and see if there's any change.

```
python -m pip install IMDBTraktSyncer --upgrade
```

If you're still having issues, see if you can provide some more detailed logs from log.txt in your install directory. That might help me narrow down where the problem is, and if there is any patterns to look for if it's breaking in the same place.

**Comment by profucius** [comment]
Hey, so unfortunately it failed again. This time the log output was much more succinct. Here is the log.txt

```
2025-01-18 23:46:06,835 - ERROR - An error occurred while running the script.
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 234, in main
    imdb_watchlist, imdb_ratings, imdb_reviews, errors_found_getting_imdb_reviews = imdbData.getImdbData(imdb_username, imdb_password, driver, directory, wait)
                                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\imdbData.py", line 65, in getImdbData
    raise PageLoadException(f"Failed to load page. Status code: {status_code}. URL: {url}")
IMDBTraktSyncer.imdbData.PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/exports/
```
I tried going to `https://www.imdb.com/exports/` from the log file, using the Chrome for Testing that the IMDBTraktSyncer script downloads. I can access it just fine.

Do you think that maybe the script is failing because it tries once to load that page, but for some reason it's not loading on the first attempt? Could making the script try a few times (with a wait in between) resolve it?

**Comment by RileyXX** [comment]
In my testing with the chrome for testing bug I was talking about earlier I noticed I got this error after downloading the IMDB exports, and then loading the next page it will hang indefinitely causing the following error: `IMDBTraktSyncer.imdbData.PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/exports/`. The script should already work around this issue by creating a new tab, loading the new page, and then closing the old ones. I think there might be something going wrong here with this temporary workaround.

To help troubleshoot, in your install directory, find the file called `checkChrome.py`, open it in a text editor, find the [line](https://github.com/RileyXX/IMDB-Trakt-Syncer/blob/765793e49ebbaa08b101c107793bd4a1f2aca563/IMDBTraktSyncer/checkChrome.py#L29) `headless = True` and change the value to `headless = False`. This will run the browser in non headless mode, so you can see exactly what the browser is doing as the script runs. Try to pay close attention to the way it is opening and closing tabs. It should open a new tab, switch to the new tab, close the old tab, and then load the new url. If you notice something different, like browser tabs not being closed, or the script not is switching browser tabs that would be good to know. 

One theory I have, possibly the script is not allowing enough time for the new tab to load, it is not switching to the new tab, or there is a chrome permissions issue preventing the browser from opening a new tab. All of which could contribute to the errors you're getting.

You could test another theory I have on fix for this by opening `errorHandling.py` file in a text editor, find the [line](https://github.com/RileyXX/IMDB-Trakt-Syncer/blob/765793e49ebbaa08b101c107793bd4a1f2aca563/IMDBTraktSyncer/errorHandling.py#L165), `driver.execute_script("window.open();")`, add a new line after it `time.sleep(5)`, then find the [line](https://github.com/RileyXX/IMDB-Trakt-Syncer/blob/765793e49ebbaa08b101c107793bd4a1f2aca563/IMDBTraktSyncer/errorHandling.py#L172) `driver.close()`, add a new line after it `time.sleep(5)`. Make sure the new lines you add have the same amount of spaces before it as the previous line to avoid a syntax error. The updated code would look like this:

````python
            # Temporary solution to Chromium bug: Restart tab logic. See: https://issues.chromium.org/issues/386887881
            # Open a new tab and close any extras from previous iterations
            driver.execute_script("window.open();")
            time.sleep(5)
            new_tab = driver.window_handles[-1]
            driver.switch_to.window(new_tab)

            # Close all other tabs except the current new tab
            for handle in driver.window_handles[:-1]:
                driver.switch_to.window(handle)
                driver.close()
                time.sleep(5)
            driver.switch_to.window(new_tab)
````

Save the files and run the script again. This change will make the browser wait 5 seconds after opening a new tab, and after closing a tab, to allow time for your it to process the actions. If the browser tab actions are happening too quickly or too slowly, it could cause these issues. Adding these waits might fix these errors you're getting, or the problem could be something else entirely. It's hard to say because I can't reproduce this error on my end, and the error logs provided don't seem to be pointing to a specific breaking point. The issue does seem to have something to do with the driver instance, which would point to issues with the chrome tabs as mentioned, browser closing prematurely, or some other related issue. I think running in non headless mode with the updated code will help better understand what's happening.

**Comment by profucius** [comment]
Very detailed and helpful response. So first I tried the headless=False method and watched the script run. Btw, that was fun. I do automations of other kinds myself, so seeing a website automation run like this gives me all sorts of ideas.

Anyway the tabs loaded fine, especially the Export one, and it seemed to run fine for a while. The export was In Progress for a while and it refreshed occasionally looking for that to finish. Once it did it downloaded the csv and proceeded. It loaded a few manual IDMB pages, and then stopped. Then the log output dumped errors again. Note, the below is without increasing the sleep time to 5 seconds.

```
C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer>IMDBTraktSyncer
Starting IMDBTraktSyncer....
Your settings are saved at:
D:\Apps\[AppData]\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer
Starting WebDriver...
Successfully signed in to IMDB
Processing Trakt Data
Processing Trakt Data Complete
Processing IMDB Data
Processing IMDB Data Complete
   - Selenium WebDriver Error: Message: invalid session id: session deleted as the browser has closed the connection URL: https://www.imdb.com/title/tt5685404/ (File: errorHandling.py, Line: 231)
   - Non-retryable WebDriver error encountered. Aborting.
Error resolving IMDB_ID tt5685404: Failed to load page. Status code: None. URL: https://www.imdb.com/title/tt5685404/
   - Selenium WebDriver Error: Message: invalid session id URL: https://www.imdb.com/title/tt4259698/ (File: errorHandling.py, Line: 231)
   - Non-retryable WebDriver error encountered. Aborting.
```
followed by a lot of individual IMDB page records like the above.
and then the same error logs by the end that I've quoted prior.

(splitting my two attempts into two comments, so you can process and reply to them as needed)

**Comment by profucius** [comment]
So then I ran the whole test again, `headless=False`, and with the `time.sleep(5)` included.
The very first run, it failed within the first minute, with this log:
```
2025-01-19 11:04:14,242 - ERROR - An error occurred while running the script.
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 840, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id
Stacktrace:
	GetHandleVerifier [0x00007FF7A196FBE5+28741]
	(No symbol) [0x00007FF7A18D7890]
	(No symbol) [0x00007FF7A1774E09]
	(No symbol) [0x00007FF7A17B90E4]
	(No symbol) [0x00007FF7A17EFDB9]
	(No symbol) [0x00007FF7A17EA76B]
	(No symbol) [0x00007FF7A17E9836]
	(No symbol) [0x00007FF7A1742EC5]
	GetHandleVerifier [0x00007FF7A1CAA73D+3414941]
	GetHandleVerifier [0x00007FF7A1CBE64A+3496618]
	GetHandleVerifier [0x00007FF7A1CB413D+3454365]
	GetHandleVerifier [0x00007FF7A1A3848B+850155]
	(No symbol) [0x00007FF7A18E37FF]
	(No symbol) [0x00007FF7A1741AD4]
	GetHandleVerifier [0x00007FF7A1D25198+3917304]
	BaseThreadInitThunk [0x00007FFF0C2FDBE7+23]
	RtlUserThreadStart [0x00007FFF0D065A4C+44]
```
I figured it was a fluke so I tried running again immediately after.
This time the script seems to work, it is opening tabs, refreshing the Export tab, opening individual pages, etc.
I had to walk away while it was running because Chrome kept jumping to the forefront window every time a new tab was opened.
When I came back it appears to have finished, I checked the log.txt and below was added.

```

````````````````````````````````````````````````````````````````````````````````````````````````````
2025-01-19 11:29:03,025 - ERROR - Failed to add item (28 of 143): Rick and Morty (2013) to IMDB Watchlist (tt2861424)
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 389, in main
    raise PageLoadException(f"Failed to load page. Status code: {status_code}. URL: {url}")
PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/title/tt2861424/
````````````````````````````````````````````````````````````````````````````````````````````````````

````````````````````````````````````````````````````````````````````````````````````````````````````
2025-01-19 11:29:03,029 - ERROR - Failed to add item (29 of 143): Only Murders in the Building (2021) to IMDB Watchlist (tt11691774)
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 389, in main
    raise PageLoadException(f"Failed to load page. Status code: {status_code}. URL: {url}")
PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/title/tt11691774/
````````````````````````````````````````````````````````````````````````````````````````````````````
```
many more records like this. It appears they started on 28 of 143, and continued incrementing steadily all the way to 143. So something happened on the 28th tab.
What is strange to me though is that I peaked in occassionally, and I did see IMDB content on the screen. It wasn't not fetching the data, but the log says it failed to load the page. Hmm..
Then the bottom of the log was the same (I believe) as I've quoted in previous messages in this thread.

```
2025-01-19 11:29:03,331 - ERROR - An error occurred while running the script.
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 840, in main
    driver.close()
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 495, in close
    self.execute(Command.CLOSE)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\webdriver.py", line 384, in execute
    self.error_handler.check_response(response)
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 232, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidSessionIdException: Message: invalid session id
Stacktrace:
	GetHandleVerifier [0x00007FF7A196FBE5+28741]
	(No symbol) [0x00007FF7A18D7890]
	(No symbol) [0x00007FF7A1774E09]
	(No symbol) [0x00007FF7A17B90E4]
	(No symbol) [0x00007FF7A17EFDB9]
	(No symbol) [0x00007FF7A17EA76B]
	(No symbol) [0x00007FF7A17E9836]
	(No symbol) [0x00007FF7A1742EC5]
	GetHandleVerifier [0x00007FF7A1CAA73D+3414941]
	GetHandleVerifier [0x00007FF7A1CBE64A+3496618]
	GetHandleVerifier [0x00007FF7A1CB413D+3454365]
	GetHandleVerifier [0x00007FF7A1A3848B+850155]
	(No symbol) [0x00007FF7A18E37FF]
	(No symbol) [0x00007FF7A1741AD4]
	GetHandleVerifier [0x00007FF7A1D25198+3917304]
	BaseThreadInitThunk [0x00007FFF0C2FDBE7+23]
	RtlUserThreadStart [0x00007FFF0D065A4C+44]
```

**Comment by siewers** [comment]
I don't know if this is strictly related, but I had the same error as reported.
I just deleted the entire `%USERPROFILE%\.cache\selenium` folder, and it all started working.
I have no clue why that helped, but it did.

**Comment by profucius** [comment]
> I just deleted the entire `%USERPROFILE%\.cache\selenium` folder, and it all started working.

I just tried this myself, but I am still getting errors about "invalid session id"

**Comment by siewers** [comment]
@profucius, I should also mention that I cleaned and reinstalled Python completely. I had multiple Python versions installed from different sources (Winget _and_ Chocolatey), but I removed all of them and installed Python using the official installation from the website. I'm no Python expert, but I know there can be issues with identifying the correct locations if your PATH contains multiple Python installations.
Also, the location of libraries varies depending on whether you are running Python from an elevated (admin) terminal or not.

**Comment by bequbed** [comment]
> I made a few more improvements with driver handling in [v3.3.1](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/v3.3.1).
> 
> Give it a try and see if there's any change.
> 
> ```
> python -m pip install IMDBTraktSyncer --upgrade
> ```
> 
> If you're still having issues, see if you can provide some more detailed logs from log.txt in your install directory. That might help me narrow down where the problem is, and if there is any patterns to look for if it's breaking in the same place.
> 
> Another thing you could try in the meantime, download the latest stable version of [chrome for testing](https://googlechromelabs.github.io/chrome-for-testing/) browser and chromedriver. Then replace the ones downloaded in your IMDBTraktSyncer install directory. In case the .zip extraction process is being corrupted somehow.

I would just like to say that like others I have a similar issue and for now I am able to resolve it by manually downloading chrome and chromedriver and placing it in the install directory.

**Comment by RileyXX** [comment]
Thanks for the feedback guys. It really helps when trying to reproduce and troubleshoot these errors. 

There's a few new changes in [3.4.2](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/3.4.2) which addresses handling the chrome/chromedriver downloads. 

I'm curious to know if it fixes the issue you were having @bequbed.

```
python -m pip install IMDBTraktSyncer --upgrade
```

**Comment by bequbed** [comment]
> Thanks for the feedback guys. It really helps when trying to reproduce and troubleshoot these errors.
> 
> There's a few new changes in [3.4.2](https://github.com/RileyXX/IMDB-Trakt-Syncer/releases/tag/3.4.2) which addresses handling the chrome/chromedriver downloads.
> 
> I'm curious to know if it fixes the issue you were having [@bequbed](https://github.com/bequbed).
> 
> Updating:
> 
> ```
> python -m pip install IMDBTraktSyncer --upgrade
> ```
> 
> After updating run the following to clear the chrome/chromedriver cache:
> 
> ```
> IMDBTraktSyncer --clear-cache
> ```

This is working beautifully now. I have done few runs now and no issues encountered. 

However please see attached screenshot regarding reaching the IMDB limit it seems like. 

![Image](https://github.com/user-attachments/assets/8daa6c52-344c-4ba3-a75f-936cb775952e)

**Comment by profucius** [comment]
RileyXX, I've tried the newest 3.4.2, and for some reason I am still getting the same issue. I first ran the update, then the clear cache, then ran the script. It downloaded chrome again and then started parsing the IMDB and Trakt datas. When it reached "Processing IMDB Data" it says the same thing it has been for a number of version now:

```
- Selenium WebDriver Error: Message: invalid session id: session deleted as the browser has closed the connection URL: https://www.imdb.com/exports/ (File: errorHandling.py, Line: 234)
- Non-retryable WebDriver error encountered. Aborting.
```
And then outputs the other debug lines, mentioning the noteworthy
```
File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\imdbData.py", line 118, in download_imdb_exports
    raise PageLoadException(f"Failed to load page. Status code: {status_code}. URL: {url}")
IMDBTraktSyncer.imdbData.PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/exports/
```

Is there any other way I can manually test this? I just ran it all again with the headless=false and timeout 2sec from previous messages. It is loading IMDB pages just fine, but then inevitably the macro stops with the same above errors. I can't know which page it stops on without actively watching the macro which is quite tedious. I'm assuming it's probably the /exports/ page, but again I can access that page just fine through the any browser on the same machine, including the Chrome for Testing that gets downloaded by the script.

**Comment by profucius** [comment]
Just updated to 3.4.3 after seeing the above ticket mention. Ran the script, and I think I see you also added a timed out catcher so that it doesn't error out. It seems to be working. The first few IMDB Watchlist items it was processing hung and had to retry once, but then it proceeded, and many more items processed without needing retries after that. It actually proceeded to Setting IMDB Watch History Items, which it never reached before.

It is on 29 of 5544 items, so might be waiting a while to know if it succeeded. But so far so good! It did need to retry on one of these, so this gives me hope.

One quick mention, yesterday's version was asking me the Y/N questions of if I wanted to sync Ratings etc, but today's 3.4.3 did not ask me those questions, it only asked if I wanted to remove X days from watchlist (which I said No to because I pay for Trakt). Is this a intended?

**Comment by profucius** [comment]
Uh oh, the script failed again. I had it running on my second monitor, and suddenly I saw the cmd window rapidly scrolling text. I can't even get back up to the text just as it happened because there are thousands of lines. This is the earliest one I could copy, although they all say this about "invalid session id" (which is the same error I've been having down all previous version of the script.

```
 - Failed to add item (2660 of 5544): [S02E18] Adam Ruins Everything: The First Factsgiving (2018) to IMDB Watch History (tt8215342)
   - Selenium WebDriver Error: Message: invalid session id URL: https://www.imdb.com/title/tt4537954/ (File: errorHandling.py, Line: 234)
   - Non-retryable WebDriver error encountered. Aborting.
 - Failed to add item (2661 of 5544): [S07E09] Archer: Deadly Velvet (1) (2016) to IMDB Watch History (tt4537954)
   - Selenium WebDriver Error: Message: invalid session id URL: https://www.imdb.com/title/tt4537958/ (File: errorHandling.py, Line: 234)
   - Non-retryable WebDriver error encountered. Aborting.
 - Failed to add item (2662 of 5544): [S07E10] Archer: Deadly Velvet (2) (2016) to IMDB Watch History (tt4537958)
   - Selenium WebDriver Error: Message: invalid session id URL: https://www.imdb.com/title/tt5825432/ (File: errorHandling.py, Line: 234)
   - Non-retryable WebDriver error encountered. Aborting.
```

I looked in the log.txt file, and it looks like all the errors are "Failed to load page"

```
2025-01-23 11:15:06,626 - ERROR - Failed to add item (104 of 5544): The Social Network (2010) to IMDB Watch History (tt1285016)
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 1190, in main
    raise PageLoadException(f"Failed to load page. Status code: {status_code}. URL: {url}")
PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/title/tt1285016/

2025-01-23 11:15:06,630 - ERROR - Failed to add item (105 of 5544): Source Code (2011) to IMDB Watch History (tt0945513)
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 1190, in main
    raise PageLoadException(f"Failed to load page. Status code: {status_code}. URL: {url}")
PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/title/tt0945513/

2025-01-23 11:15:06,633 - ERROR - Failed to add item (106 of 5544): Star Trek (2009) to IMDB Watch History (tt0796366)
Traceback (most recent call last):
  File "C:\Users\Admin\AppData\Local\Programs\Python\Python312\Lib\site-packages\IMDBTraktSyncer\IMDBTraktSyncer.py", line 1190, in main
    raise PageLoadException(f"Failed to load page. Status code: {status_code}. URL: {url}")
PageLoadException: Failed to load page. Status code: None. URL: https://www.imdb.com/title/tt0796366/
```
etc etc all the way to the end.

**Comment by profucius** [comment]
Just a quick update, I had some free time this weekend and wrote myself an automation to restart the script whenever the log file contains traceback errors. I have it running on a loop right now so that it processes my 5000+ backlog. My line of thinking is that if I'm the only one struggling with the issues I have, then I may as well do this, and once it's fully caught up, if I don't have anymore major issues then we can consider this resolved. I'll post again once the backlog is finished processing, it might take the rest of the day.
